man/tamper.Rd

```diff
@@ -0,0 +1,59 @@
+% Generated by roxygen2 (4.1.1): do not edit by hand
+% Please edit documentation in R/tamper.R
+\name{tamper}
+\alias{tamper}
+\title{Investigate a pipe, after an error}
+\usage{
+tamper()
+}
+\description{
+\code{tamper} is a function that can be used as an error callback,
+similarly to \code{utils::recover}. Tamper is pipe-friendly: it will
+show you the exact place of the error in the pipe.
+}
+\section{Example}{
+
+
+After running the following code: \preformatted{options(error = tamper)
+
+f <- function(data) {
+   data \%>\%
+     (function(x) force(x)) \%>\%
+     multiply_by(10) \%>\%
+     add(10) \%>\%
+     add("oh no!") \%>\%
+     subtract(5) \%>\%
+     divide_by(5)
+}
+
+ 1:10 \%>\%
+   multiply_by(2) \%>\%
+   f() \%>\%
+   add(1:10)
+}
+you will see: \preformatted{Error in add(., "oh no!") : non-numeric argument to binary operator
+
+Enter a pipe stage number, 1 to switch mode, or 0 to exit
```

Comment:
So if you have 

```r
library(magrittr)
options(error = recover)
1:5 %>% multiply_by(10) %>% add("oh no!") %>% subtract(5)
```
then you'll see 

```r
Error in add(., "oh no!") : non-numeric argument to binary operator

Enter a frame number, or 0 to exit

 1: 1:5 %>% multiply_by(10) %>% add("oh no!") %>% subtract(5)
 2: withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
 3: eval(quote(`_fseq`(`_lhs`)), env, env)
 4: eval(expr, envir, enclos)
 5: `_fseq`(`_lhs`)
 6: freduce(value, `_function_list`)
 7: Recall(function_list[[1]](value), function_list[-1])
 8: freduce(value, `_function_list`)
 9: Recall(function_list[[1]](value), function_list[-1])
10: freduce(value, `_function_list`)
11: withVisible(function_list[[1]](value))
12: function_list[[1]](value)
13: function_list[[1]](value)
```

The error happens in **10**. **8** is still OK, you can go there and check `value`. In **10** you already get `promise already under evaluation`, because it was not possible to evaluate `value` there. The actual `add` is not listed in the stack, I am not sure why, actualy.